Convert 'testing faq' to html #52

crispyfridge opened the pull request:
Fixes #52 
Link to preview: http://crispyfridge.github.io/website/contents/testingFaq.html

damithc commented on contents/testingFaq.html:
```diff
@@ -0,0 +1,114 @@
+    String content = getFileContent(fileName);<br>
+         print(content);<br>
+  }</code>
+</div>
```
In the code examples above, indentation is not correct, in the source code as well as when displayed.

damithc commented:
@crispyfridge After this one, move on to harder ones involving more css and/or JS, leaving easier ones to new comers.

damithc commented:
I wonder if we can use Gist to embed formatted code snippets. Can you explore that?

crispyfridge commented:
@damithc Yeah, gists can be embedded in html files. The code snippet would look like the following:
![image](https://cloud.githubusercontent.com/assets/7552278/11612929/b4b0d33c-9c48-11e5-9280-6c5dcb548971.png)
However, the gist created will be hosted on my site, and can only be edited by me. 
Alternatively, I could create an anonymous gist, but an anonymous gist cannot be edited and deleted.

damithc commented:
It would be inconvenient if the code is in a different repo. See if there is any way to serve formatted code from our own repo. 
Also explore other ways of embedding formatted code. If we figure out a way, we can use it in other places too.

crispyfridge commented:
There is a way to display formatted code from github in html using jsfiddle. However, there is it rather inconvenient, as a specific folder has to be created for the code to be shared, and the files have to be titled the same way, i.e `<title>.html/js/css`. The jsfiddle code display also includes tabs such as css and javascript, which cannot be modified.

The suggested method to showing formatted code from w3schools is wrapping the code in  `<pre>` and `<code>` tags, and using google/code-prettify script with the tags colours the formatted code.

wkurniawan07 commented:
I don't think it needs a complicated solution. Just wrapping it with `<pre>` preserves the whitespace, and everything else is just a matter of styling.

damithc commented:
For this particular instance we don't need anything more complicated than `<pre>` tags. But I'm interested in finding a general solution for showing code with automatic syntax coloring. 
@crispyfridge can you point to an example of the 'suggested method' you have mentioned? Does it have automatic syntax coloring?

crispyfridge commented:
@damithc Yes it has syntax coloring. This is an example page using google/code-prettify:
 https://rawgit.com/google/code-prettify/master/examples/quine.html
It uses the script tag`<script src="https://cdn.rawgit.com/google/code-prettify/master/loader/run_prettify.js"></script>` and including the class prettyprint with the language of the code snippet, within the `pre` tag that wraps the code, like `<pre class="prettyprint lang-html">`
More can be found on https://github.com/google/code-prettify

crispyfridge commented:
I've added the use of `<pre>` and `<code>` for displaying the formatted code

damithc commented:
The code-pretty thing looks promising. Shall we try it in this PR?

crispyfridge commented:
I've added the code-prettify in the code blocks.
Preview: http://crispyfridge.github.io/website/contents/testingFaq.html

damithc commented on contents/testingFaq.html:
```diff
@@ -0,0 +1,108 @@
+It can be refactored like this:
+<pre class="prettyprint"><code class="language-java">void display(){
+    String content = getFileContent(fileName);
+         print(content);
```
Indentation problem?

damithc replied:
Why this line is indented two levels. It should be the same as the previous line right?

damithc commented on contents/testingFaq.html:
```diff
@@ -0,0 +1,108 @@
+}
+</code></pre>
+It can be refactored like this:
+<pre class="prettyprint"><code class="language-java">void display(){
```
I guess adding a line break just before void makes the html file neater but adds an extra line in the display view?

damithc commented:
Nice. Does it support a lighter theme?

crispyfridge commented:
@damithc Adding a line break just before void adds an extra line in the display view. Should I add the line break to make the html file neater?
code-prettify only has one theme where the background is not black, but the overall look is not as great.
Preview: http://crispyfridge.github.io/website/contents/testingFaq.html

damithc commented:
OK, no need to add a line break. We can stick with the dark theme.
Does it allow formatting a word that occurs inside a paragraph? e.g. a mention of a class/method name?
The only other pending change is the indentation problem mentioned earlier.

damithc commented:
Looks like it can format words inside a paragraph. Can you apply it to words inside paragraphs that represent code elements, the way you've done it for `CityConnnect.java`?

crispyfridge commented:
I didn't apply syntax highlighting to `CityConnect.java`, but wrapped the `<code>` tags around it. I can do that for the other words that represent code elements. 

For the indentation, I'm not very sure...Is it better if the `<pre>` tag follows the sentenence '...the following method' without the line break? Or should I give an extra indent to the contents within the `<div>`?

damithc commented:
Sure, just `<code>` wrapping is enough.
As for indentation, I'm referring to extra indentation in line 50. Added another comment there.

crispyfridge commented:
Line 50 was formatted that way in the google docs, so I thought it was meant to be that way. :sweat_smile: 
Fixed the indentation for that line, and added the `<code>` tags to code elements in paragraph. Also added some urls that were missing from the html.
Preview: http://crispyfridge.github.io/website/contents/testingFaq.html

damithc commented:
@wkurniawan07 any further comments?

wkurniawan07 commented:
Just some coding style improvements:

1. `void display() {` instead of `void display(){`
2. `// Comment` instead of `//Comment` (not entirely sure if there's an actual convention on this, but most of the Java codes I see have space in that place)
3. Missing space between `>` and `output` in `TextBuddy.exe < input.txt >output.txt`

Everything else looks ok, I like this library :+1:

crispyfridge commented:
I've made the improvements.

wkurniawan07 commented on contents/testingFaq.html:
```diff
@@ -0,0 +1,109 @@
+         // display line
+}</code></pre>
+It can be refactored like this:
+<pre class="prettyprint"><code class="language-java">void display(){
```
Errrr... actually there's still one more here... :stuck_out_tongue:
